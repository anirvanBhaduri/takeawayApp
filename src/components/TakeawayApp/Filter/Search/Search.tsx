import * as React from 'react';

import {Filter, FilterProps} from "../Filter";
import {RestaurantProps} from "../../RestaurantList/Restaurant/Restaurant";

/**
 * The searchTerm is required to ensure
 * we can search.
 *
 * @interface SearchProps
 */
interface SearchProps extends FilterProps {
    searchTerm: string;
}

/**
 * Search using a string.
 *
 * Used to filter RestaurantProps.
 *
 * @class Search
 * @package TakeawayApp
 *
 * @author Anirvan Bhaduri
 * @since 9th Dec 2018
 * @version 0.0.1
 */
export class Search extends Filter<SearchProps> {

    /**
     * The search term used to filter the results.
     */
    searchTerm: string;

    /**
     * Search constructor.
     *
     * @param props
     */
    constructor(props: any) {
        super(props);

        this.searchTerm = this.props.searchTerm;
    }

    /**
     * Set the search term.
     *
     * @param searchTerm
     */
    setSearchTerm(searchTerm: string) {
        this.searchTerm = searchTerm;
    }

    /**
     * Get the current search term.
     *
     * @return string
     */
    getSearchTerm(): string {
        return this.searchTerm;
    }

    /**
     * Apply the search filter.
     *
     * @param list
     *
     * @return RestaurantProps
     */
    applyFilter(list: RestaurantProps[]): RestaurantProps[] {
        // Do some filter logic here using searchTerm
        // ... should mainly remove elements

        // When we don't have a search term, we don't want to filter at all
        if (!this.searchTerm) {
            return list;
        }

        const searchTerm = this.searchTerm.toLowerCase();
        return list.filter(
            (value: RestaurantProps): boolean => {
                return value.name.toLowerCase().indexOf(searchTerm) >= 0;
            }
        );
    }

    /**
     * Render the search component.
     *
     * @return JSX
     */
    render() {
        const handler = this.props.filterHandler;
        const searchTerm = this.props.searchTerm;

        return (
            <div>
                <input
                    onChange={handler}
                    value={searchTerm}
                    type="input"
                />
            </div>
        );
    }
}
