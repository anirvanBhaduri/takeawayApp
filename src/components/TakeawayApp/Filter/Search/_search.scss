/// Search component styles.
///
/// @author Anirvan Bhaduri
/// @since 10th Dec 2018

@import '../../../UI/_colour_constants.scss';
@import '../../../UI/size_constants';

// The search bar's styles
.search {
  float: left;
  width: 550px;

  @media screen and (max-width: $screen-break-medium) {
    width: 98%;
    margin: 10px;
  }

  @media screen and (max-width: $screen-break-small) {
    width: 98%;
    margin: 10px;
  }
}

// The search bar's input's styles
.search__input {
  display: inline-block;
  @include input-styles(28px, 545px);

  @media screen and (max-width: $screen-break-medium) {
    font-size: inherit;
  }

  @media screen and (max-width: $screen-break-small) {
    font-size: inherit;
  }
}