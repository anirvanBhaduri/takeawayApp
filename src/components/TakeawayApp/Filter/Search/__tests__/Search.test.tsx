import * as React from 'react';
import {shallow} from "enzyme";

// SUT
import {Search} from "../Search";

/**
 * Expected data that is pre sorted. Output from
 * test should result in this expected data.
 */
export const sortedData = [
    {
        favourite: false,
        name: 'what',
        status: 'open',
    }
];

/**
 * Search component tests.
 *
 * @covers Search
 */
describe('Search Component Tests', () => {

    /**
     * Apply the Search filter.
     *
     * @covers Search.applyFilter
     */
    it('should apply the search filter to the restaurant props array', () => {
        const search = new Search({
            searchTerm: 'wh',
            filterHandler: () => {}
        });

        const sorted = search.applyFilter([
            {
                favourite: false,
                name: 'what',
                status: 'open',
                clickHandler: () => {},
            },
            {
                favourite: true,
                name: 'test',
                status: 'closed',
                clickHandler: () => {},
            }
        ]);

        // Deep copy check using convoluted cloning logic.
        // I think we need to do this because the objects themselves have interesting
        // props being added to them somehow... perhaps a side effect of
        // using typescript?
        expect(JSON.parse(JSON.stringify(sorted))).toEqual(sortedData);
    });

    /**
     * Search renders correctly.
     *
     * @covers Search.render
     */
    it('should render correct snapshot of Search Component', () => {
        const wrapper = shallow(
            <Search searchTerm={'at'} filterHandler={() => {}} />
        );

        expect(wrapper).toMatchSnapshot();
    });
});
