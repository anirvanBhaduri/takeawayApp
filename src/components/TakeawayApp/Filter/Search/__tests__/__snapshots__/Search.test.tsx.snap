// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Component Tests should render correct snapshot of Search Component 1`] = `
<div
  className="search"
>
  <input
    className="search__input"
    onChange={[Function]}
    placeholder="Search"
    type="input"
    value="at"
  />
</div>
`;
